osu.Game.Rulesets.Osu/Difficulty/MathUtil/HitProbabilities.cs

```diff
@@ -0,0 +1,165 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+using osu.Game.Rulesets.Osu.Difficulty.Preprocessing;
+
+namespace osu.Game.Rulesets.Osu.Difficulty.MathUtil
+{
+    public class HitProbabilities
+    {
+        private static int cacheHit = 0;
+        private static int cacheMiss = 0;
+
+        private List<MapSectionCache> sections = new List<MapSectionCache>();
+
+        public HitProbabilities(List<OsuMovement> movements, double cheeseLevel, int difficultyCount = 20)
+        {
+            for (int i = 0; i < difficultyCount; ++i)
+            {
+                int start = movements.Count * i / difficultyCount;
+                int end = movements.Count * (i + 1) / difficultyCount - 1;
+                sections.Add(new MapSectionCache(movements.GetRange(start, end - start + 1), cheeseLevel));
+            }
+        }
+
+        public int Count(int start, int sectionCount)
+        {
+            int count = 0;
+            for (int i = start; i != start + sectionCount; i++)
+            {
+                count += sections[i].Movements.Count;
+            }
+
+            return count;
+        }
+
+        public bool IsEmpty(int sectionCount)
+        {
+            bool isEmpty = false;
+            for (int i = 0; i <= sections.Count - sectionCount; i++)
+            {
+                isEmpty = isEmpty || Count(i, sectionCount) == 0;
```

Comment:
This should just return false as soon as a non-empty section is found rather than iterating through no matter what